libpromises/rlist.c

```diff
@@ -325,18 +325,56 @@ Rlist *RlistAppendRval(Rlist **start, Rval rval)
     return rp;
 }
 
-Rval RvalNew(const void *item, RvalType type)
+Rval RvalNewRewriter(const void *item, RvalType type, JsonElement *map)
 {
     switch (type)
     {
     case RVAL_TYPE_SCALAR:
-        return (Rval) { xstrdup(item), RVAL_TYPE_SCALAR };
+        if (NULL != map && JsonLength(map) > 0)
+        {
+            // TODO: replace with BufferSearchAndReplace when the
+            // string_replace code is merged.
+            // Sorry about the CF_BUFSIZE ugliness.
+            int max_size = 10*CF_BUFSIZE+1;
+            const char* buffer_from = malloc(max_size);
+            const char* buffer_to = malloc(max_size);
+
+            Buffer *format = BufferNew();
+            strncpy(buffer_from, item, max_size);
+
+            JsonIterator iter = JsonIteratorInit(map);
+            const char *key;
+
+            for (int loop=0; loop < 3; loop++)
+            {
+                while ((key = JsonIteratorNextKey(&iter)))
+                {
+                    BufferPrintf(format, "$(%s)", key);
+                    ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+                    strncpy(buffer_from, buffer_to, max_size);
+                    BufferPrintf(format, "${%s}", key);
+                    ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+                }
+            }
```

Comment:
Sorry, but I don't think this is a good approach, because it is incredibly inefficient, and this is going to get evaluated a lot. In a body with say, 4 arguments, you will be scanning the string 3 * 4 * 2 = 24 times, even when there is not a single replacement. What about something like this (untested):

```
for (int iteration = 0; iteration < 10; iteration++)
{
    bool replacement_made = false;
    int var_start = -1;
    for (int c = 0; c < buffer_from[c]; c++)
    {
       if (buffer_from[c] == '$')
       {
          c++;
          if (buffer_from[c] == '(' || buffer_from[c] == '{')
          {
             var_start = c-1;
          }
       }
       else if (var_start >= 0 && buffer_from[c] == ')' || buffer_from[c] == '}')
       {
          // Before the replacement.
          memcpy(buffer_to, buffer_from, var_start);

          // The actual replacement.
          char saved = buffer_from[c];
          buffer_from[c] = '\0';
          const char *repl = JsonObjectGetAsString(map, buffer_from[var_start+2]);
          int repl_len = strlen(repl);
          buffer_from[c] = saved;
          memcpy(buffer_to + var_start, repl, repl_len);

          // The text after.
          strlcpy(buffer_to + var_start + repl_len, buffer_from + c + 1, max_size - var_start - repl_len);

          // Reset location to immediately after the replacement.
          c = var_start + repl_len - 1;
          var_start = -1;
          strcpy(buffer_from, buffer_to);
          replacement_made = true;
       }
    }

    if (!replacement_made)
    {
        break;
    }
}
```

I'm not taking into account that the string can look like ``$(x}`` or ``${x)``, but this is just a suggestion.